### DISCUSSION TITLE
How to create a custom handler that keeps nativ control mappings

### DISCUSSION THREAD

**Discussion by OliverDevsCodex** [dotnet/maui#17294] [Q&A] [answered]
Hey Guys,

I created a custom handler for a custom Entry control that I made to enable the possibility to enable/disable the underline in an Entry. I know I could just attach some functionality to the already existing handler, but that doesn't work with the XAML hot reload and is kind of annoying.

So far everything works fine, but now all the default functionality is gone, like setting the font size, attributes, etc. I cannot change anything anymore (duo to the custom handler I guess). Remapping every single property by myself would be insane.

How can I make a custom handler that builds on top of the basic/native functionality that already is there?

For Example, that is my control:
```
public class ExtendedEntry : Entry
{
    public static readonly BindableProperty NoUnderlineProperty =
        BindableProperty.Create(
            nameof(NoUnderline),
            typeof(bool),
            typeof(ExtendedEntry));

    public bool NoUnderline
    {
        get => (bool)GetValue(NoUnderlineProperty);
        set => SetValue(NoUnderlineProperty, value);
    }
}
```
That is the Handler

```
public partial class ExtendedEntryHandler 
{

    static IPropertyMapper<ExtendedEntry, ExtendedEntryHandler> PropertyMapper = new PropertyMapper<ExtendedEntry, ExtendedEntryHandler>(ViewHandler.ViewMapper)
    {
        [nameof(ExtendedEntry.NoUnderline)] = MapNoUnderlineProperty,
    };

    public ExtendedEntryHandler() : base(PropertyMapper)
    {
    }
}
```
And that is the Handler implementation for Android:

```
public partial class ExtendedEntryHandler : ViewHandler<ExtendedEntry, Android.Widget.EditText>
{
    
    protected override EditText CreatePlatformView() => new EditText(Context);

    protected override void ConnectHandler(EditText platformView)
    {
        base.ConnectHandler(platformView);

        // Perform any control setup here
    }

    protected override void DisconnectHandler(EditText platformView)
    {
        platformView.Dispose();
        base.DisconnectHandler(platformView);
    }

    static void MapNoUnderlineProperty(ExtendedEntryHandler handler, ExtendedEntry entry)
    {
        if (entry.NoUnderline)
            handler.PlatformView.BackgroundTintList = Android.Content.Res.ColorStateList.ValueOf(Colors.Transparent.ToAndroid());
        else
            handler.PlatformView.BackgroundTintList = Android.Content.Res.ColorStateList.ValueOf(Colors.Gray.ToAndroid());
    }
}
```
Now I can remove the underline like intended, but if I set the font size, for example, nothing happens:
![image](https://github.com/dotnet/maui/assets/101668596/fd4d3293-0bc0-414b-8e9f-fad327472c47)

Here I can set the size to 2000 and it would not change a thing:
`<controls:ExtendedEntry NoUnderline="True" Text="Test" FontSize="2000" FontAttributes="Bold"/>`

**Comment by xtuzy** [comment]
In this class
```
public partial class ExtendedEntryHandler 
{

    static IPropertyMapper<ExtendedEntry, ExtendedEntryHandler> PropertyMapper = new PropertyMapper<ExtendedEntry, ExtendedEntryHandler>(ViewHandler.ViewMapper)
    {
        [nameof(ExtendedEntry.NoUnderline)] = MapNoUnderlineProperty,
    };

    public ExtendedEntryHandler() : base(PropertyMapper)
    {
    }
}
```
Try change ViewHandler.ViewMapper to EntryHandler.ViewMapper

**Reply by OliverDevsCodex** [reply]
sadly it has no effect. still, olny the mapped property works. I feel like the passed-in IPropertyMapper is not even used...

**Comment by xtuzy** [comment, marked as answer]
your handler also need extend from EntryHandler, see https://github.com/dotnet/maui/issues/16178

**Reply by OliverDevsCodex** [reply]
thank you very much, that was it! I think i got how it works now, again thank you very much! 

For the people who stumble across this: 

I changed the class ExtendedEntryHandler (Android) to: 
```
public partial class ExtendedEntryHandler : EntryHandler//ViewHandler<ExtendedEntry, Android.Widget.EditText>
{
    static void MapNoUnderlineProperty(ExtendedEntryHandler handler, ExtendedEntry entry)
    {
        if (entry.NoUnderline)
            handler.PlatformView.BackgroundTintList = Android.Content.Res.ColorStateList.ValueOf(Colors.Transparent.ToAndroid());
        else
            handler.PlatformView.BackgroundTintList = Android.Content.Res.ColorStateList.ValueOf(Colors.Gray.ToAndroid());
    }
}
```

And also the ViewHandler.ViewMapper to EntryHandler.Mapper.

**Comment by thomasgalliker** [comment]
Nice example. My open question is: When would `DisconnectHandler` be called?
- Do I have to call it when ever the page that contains Entries is no longer part of the navigation stack?
- Does MAUI call DisconnectHandler at one point?

If I subscribe to events in ConnectHandler, I want to be sure they are unsubscribed when the control is no longer used. Otherwise, this leads to memory leaks.
